Spark-3.5: Ensure that partition stats files are considered for GC procedures

ajantha-bhat opened the pull request:
Currently depends on https://github.com/apache/iceberg/pull/8502/

Please refer only the top commit for this PR related changes.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:
```diff
@@ -148,12 +148,49 @@ public static List<String> statisticsFilesLocations(Table table) {
    * @return the location of statistics files
+   * @deprecated use the {@code statisticsFilesLocationsForSnapshots(table, snapshotIds)} instead.
    */
+  @Deprecated
```
can't use the `Predicate<StatisticsFile>` for partition stats. Hence, deprecated and introduced snapshot id based input like other methods in this class (manifest list, manifest etc)

ajantha-bhat commented on spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestExpireSnapshotsProcedure.java:
```diff
@@ -475,18 +474,9 @@ public void testExpireSnapshotsWithStatisticFiles() throws Exception {
     Assertions.assertThat(output.get(0)[5]).as("should be 1 deleted statistics file").isEqualTo(1L);
 
     table.refresh();
-    List<StatisticsFile> statsWithSnapshotId1 =
```
Since we are checking `containsExactly` at line 482, no need of these checks.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:
```diff
@@ -148,12 +148,49 @@ public static List<String> statisticsFilesLocations(Table table) {
    * @param table table for which statistics files needs to be listed
    * @param predicate predicate for filtering the statistics files
    * @return the location of statistics files
+   * @deprecated use the {@code statisticsFilesLocationsForSnapshots(table, snapshotIds)} instead.
```
Shall we add a note on when it was deprecated (1.5.0) and when it will be removed (1.6.0)?

ajantha-bhat replied:
I have added this already in previous commit.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:
```diff
@@ -148,12 +148,49 @@ public static List<String> statisticsFilesLocations(Table table) {
+   *     statistics files for all the snapshots will be returned.
+   * @return the location of statistics files
+   */
+  public static List<String> statisticsFilesLocationsForSnapshots(
```
What do you think about calling it `statisticsFilesLocations` to match `manifestListLocations` that also accepts a set of snapshot IDs?

ajantha-bhat replied:
The deprecated class already has a name `statisticsFilesLocations`. So, If I keep the same name, compiler says ambiguous error when I call `statisticsFilesLocationsForSnapshots(table, null)`

Hence, I had to keep different name.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:
```diff
@@ -148,12 +148,49 @@ public static List<String> statisticsFilesLocations(Table table) {
+  public static List<String> statisticsFilesLocationsForSnapshots(
+      Table table, Set<Long> snapshotIds) {
+    List<String> statsFileLocations = Lists.newArrayList();
+    Predicate<StatisticsFile> statsFilePredicate;
```
What about an empty line before predicates to separate this block?

aokolnychyi commented on core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:
```diff
@@ -148,12 +148,49 @@ public static List<String> statisticsFilesLocations(Table table) {
+    List<String> statsFileLocations = Lists.newArrayList();
+    Predicate<StatisticsFile> statsFilePredicate;
+    Predicate<PartitionStatisticsFile> partitionStatsFilePredicate;
+    if (snapshotIds == null) {
```
What about the following to shorten the vars and avoid splitting the logic into multiple lines?

```
Predicate<StatisticsFile> statsFilePredicate;
Predicate<PartitionStatisticsFile> partitionStatsFilePredicate;
if (snapshotIds == null) {
  statsFilePredicate = file -> true;
  partitionStatsFilePredicate = file -> true;
} else {
  statsFilePredicate = file -> snapshotIds.contains(file.snapshotId());
  partitionStatsFilePredicate = file -> snapshotIds.contains(file.snapshotId());
}
```

aokolnychyi commented:
@ajantha-bhat, this looks mostly good to me. Can we add the tests to all Spark versions to be safe?

ajantha-bhat commented:
> @ajantha-bhat, this looks mostly good to me. Can we add the tests to all Spark versions to be safe?

Sure. I didn't initially add to other spark version thinking if I get any comments for test, I have to rework in all the spark version. So, I thought of adding tests in a single follow up PR for other spark versions. 

Now that we don't have any specific comment on test, I can add it in this PR itself.

ajantha-bhat commented:
@aokolnychyi: Thanks for the review. 
Addressed all the comments except https://github.com/apache/iceberg/pull/9284#discussion_r1432485284 (with a reasoning). 

Build passed and PR is ready.

ajantha-bhat commented:
ping @aokolnychyi

ajantha-bhat commented:
> It looks good to me, I had one optional comment. Let me know what you want to do there, @ajantha-bhat.

@aokolnychyi : Are you referring to https://github.com/apache/iceberg/pull/9284#discussion_r1432485284 comment? I have replied why I can't change the name due to compile time ambiguity of same signature.
Other than that I don't see any more open comment. So, please let me know.

aokolnychyi commented:
Oops, I probably forgot to post the comment. I was thinking about `statisticsFileLocations` to match the naming of other methods like `metadataFileLocations` (`Files` -> `File`). It is optional, though.

ajantha-bhat commented:
> Oops, I probably forgot to post the comment. I was thinking about statisticsFileLocations to match the naming of other methods like metadataFileLocations (Files -> File). It is optional, though.

`statisticsFileLocations` is an existing public method (before this PR). So, we can't rename it directly. We have to deprecate and introduce new one. I think we can leave it as it as. 

The new method `statisticsFilesLocationsForSnapshots` can be renamed but it will become inconsistent with existing `statisticsFileLocations`.

aokolnychyi commented:
Yep, that's fair. Let's keep it as is then.
